programs/lz4io.c

```diff
@@ -395,8 +404,50 @@ typedef struct {
     void*  dstBuffer;
     size_t dstBufferSize;
     LZ4F_compressionContext_t ctx;
+    LZ4F_CDict* cdict;
 } cRess_t;
 
+static void LZ4IO_loadDict(char **dictionaryBuffer, size_t *dictionarySize) {
+    FILE* dictionaryFile;
+    size_t blockSize = 64 KB;
+    size_t dictionaryBufferSize = blockSize;
+    size_t readSize;
+    *dictionarySize = 0;
+    *dictionaryBuffer = malloc(dictionaryBufferSize);
+
+    if (!g_dictionaryFilename) {
+        EXM_THROW(25, "Dictionary error : no filename provided");
+    }
+
+    dictionaryFile = LZ4IO_openSrcFile(g_dictionaryFilename);
+
+    do {
+        if (*dictionarySize + blockSize > dictionaryBufferSize) {
+            dictionaryBufferSize *= 2;
+            *dictionaryBuffer = realloc(*dictionaryBuffer, dictionaryBufferSize);
```

Comment:
`realloc()` is a pretty capricious function.
If it fails, `*dictionaryBuffer == NULL`, but buffer previously accessed through `*dictionaryBuffer` is still valid, but cannot be accessed anymore (since we just lost its pointer). So it cannot be freed, so it becomes a memory leak.

In this case though, it does not make a big different, because you `EXM_THROW()` right afterwards, which terminates the program. So no harm done.

Just keep in mind that using `realloc()` can be relatively tricky. Avoid it when not necessary.

Reply:
I meant to ask, what is meant by the number in the `EXM_THROW` macro?

Reply:
This is the error code that the program will return (`exit(errorCode)`).
It will also be displayed in the error message.
It's a cheap way to find were the error was triggered, especially when 2 error messages are similar.